### PR TITLE
[WIP] Add HSDP dtensor state_dict support

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105990

This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.
4. Utilize the _device_mesh to do operations for state_dict and optim_state_dict manipulation.

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_dtensor_state_dict.py [hunk 1]
```diff
@@ -227,6 +228,30 @@ def test_dtensor_sharded_model_load_state_dict(self, offload_to_cpu):
...
+    def test_fsdp_device_mesh_init_2d(self):
+        mesh_tensor = torch.arange(self.world_size).view(2, -1)
+        device_mesh = DeviceMesh(self.device_type, mesh_tensor)
+        model = FSDP(
+            TestDummyModel().cuda(),
+            _device_mesh=device_mesh,
+            sharding_strategy=ShardingStrategy.HYBRID_SHARD,
+        )
+        optim = torch.optim.Adam(model.parameters(), lr=0.1)
+        model(model.get_input()).sum().backward()
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:252]: nit (i know the PR is WIP) ideally the test should validate something like we're sharded / not sharded over the right dimensions?

File: torch/distributed/fsdp/_init_utils.py [hunk 2]
```diff
@@ -98,21 +99,34 @@ def _init_process_group_state(
     process_group: ProcessGroupType,
     sharding_strategy: ShardingStrategy,
     policy: Optional[_FSDPPolicy],
+    device_mesh: DeviceMesh = None,
 ) -> _FSDPState:
+    if process_group is not None and device_mesh is not None:
+        raise ValueError(
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_init_utils.py:105]: nit: add a test for this

**Review by rohan-varma — COMMENTED** [review]
per discussion with @fegin, it might be important for PT-D to ensure composable API stays in sync with FSDP as a first-class citizen. Might be good to sync w/ @fegin to see if we should include composable API changes in this PR as well.

File: test/distributed/fsdp/test_fsdp_dtensor_state_dict.py [hunk 3]
```diff
@@ -227,6 +228,30 @@ def test_dtensor_sharded_model_load_state_dict(self, offload_to_cpu):
...
+
+    @with_comms
+    @skip_if_lt_x_gpu(2)
+    def test_fsdp_device_mesh_init_2d(self):
+        mesh_tensor = torch.arange(self.world_size).view(2, -1)
+        device_mesh = DeviceMesh(self.device_type, mesh_tensor)
+        model = FSDP(
+            TestDummyModel().cuda(),
+            _device_mesh=device_mesh,
```
[the remark this answers is not in the document]
Comment (wz337) [reply test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:248]: I am hardcoding the mesh dim for HSDP sharded and replicate for now, since we want to quickly roll this part out so modelstore folks can start build the integration for sharded DTensor as well 2D device-mesh support. 

For users, they actually don't have to know the mesh dimension for dp replicate and dp shard, since the default would be dp_replicate: 0, dp_shard: 1. For more advanced users, they can pass an additional arguments in initializing device_mesh. More concrete example is in this diff Chien-Chin has. D47781455

**Comment by wz337** [comment]
close as we have https://github.com/pytorch/pytorch/pull/107533
